File: companion-docker/app/rmc_handshake.py
"""RMC platform handshake (v3.37.0 Agent 4 — Docker sibling).

TARGET: RunMyCampus's OWN ``/api/v1/...`` endpoints. This module
NEVER programmatically logs into any third-party SIS. Vendor data
extraction lives in ``companion-extension/`` (the browser extension
where the operator's own authenticated tab is the security boundary).
The Docker sibling's only login is to the RMC platform itself.

Mirrors the Tauri sibling's `rmc_handshake.rs` so both siblings keep
identical operator UX.
"""
from __future__ import annotations

import dataclasses
import hmac
import logging
from typing import Optional

import httpx

logger = logging.getLogger(__name__)

RMC_LOGIN_PATH = "/api/v1/auth/login/"
RMC_MAA_TEXT_PATH = "/api/v1/migration/maa/text/"
RMC_MAA_SIGN_PATH = "/api/v1/migration/maa/sign/"
RMC_RECEIPT_PATH_PREFIX = "/api/v1/migration/receipts/"

DEFAULT_TIMEOUT_SECS = 30.0


class HandshakeError(Exception):
    """Typed handshake error. Subclassed below for known cases."""


class BadServerUrl(HandshakeError):
    pass


class TransportError(HandshakeError):
    pass


class Unauthorized(HandshakeError):
    """HTTP 401 from RMC platform login."""


class Forbidden(HandshakeError):
    pass


class NotFound(HandshakeError):
    pass


class BadResponse(HandshakeError):
    pass


class MaaSignRejected(HandshakeError):
    pass


@dataclasses.dataclass
class MaaText:
    vendor: str
    version: str
    text: str
    is_draft: bool
    active_version: str


@dataclasses.dataclass
class MAAReceipt:
    maa_id: int
    signed_at: str
    agreement_version: str


@dataclasses.dataclass
class Receipt:
    receipt_id: int
    bundle_id: Optional[int]
    status: str


def _join_url(base: str, path: str) -> str:
    if not (base.startswith("http://") or base.startswith("https://")):
        raise BadServerUrl("server URL must include http:// or https://")
    return f"{base.rstrip('/')}{path}"


def _classify_status(status: int) -> HandshakeError:
    if status == 401:
        return Unauthorized(f"authentication failed status={status}")
    if status == 403:
        return Forbidden(f"forbidden status={status}")
    if status == 404:
        return NotFound(f"not found status={status}")
    return HandshakeError(f"unexpected status={status}")


def constant_time_text_eq(a: str, b: str) -> bool:
    """Constant-time equality for MAA signature_text echo + tokens.

    Python has no native zeroize; we use ``hmac.compare_digest`` which
    is the stdlib's constant-time primitive.
    """
    return hmac.compare_digest(a.encode("utf-8"), b.encode("utf-8"))


class RMCSession:
    """Holds the bearer token in memory only. The token is never written
    to disk, never logged, never echoed to stdout.
    """

    def __init__(
        self,
        server_url: str,
        *,
        client_factory=None,
    ) -> None:
        self.server_url = server_url.rstrip("/")
        # Caller may inject a httpx MockTransport for tests.
        self._client_factory = client_factory or (
            lambda: httpx.Client(verify=True, timeout=DEFAULT_TIMEOUT_SECS)
        )
        self._token: Optional[str] = None

    @property
    def token(self) -> Optional[str]:
        return self._token

    def clear_token(self) -> None:
        # Best-effort overwrite; Python strings are immutable so the
        # GC handles real destruction. Drop the reference.
        if self._token is not None:
            self._token = "0" * len(self._token)
            self._token = None

    def _auth_header(self) -> dict[str, str]:
        if not self._token:
            return {}
        return {"Authorization": f"Bearer {self._token}"}

    def login(self, email: str, password: str) -> str:
        """POST /api/v1/auth/login/ — returns bearer token."""
        url = _join_url(self.server_url, RMC_LOGIN_PATH)
        try:
            with self._client_factory() as client:
                response = client.post(url, json={"email": email, "password": password})
        except httpx.HTTPError as exc:
            raise TransportError(f"transport error: {type(exc).__name__}") from exc
        # Best-effort overwrite the password var locally. (Caller should
        # del password after this returns.)
        password = "0" * len(password)
        del password
        status = response.status_code
        logger.info("rmc_handshake.login status=%s", status)
        if status >= 400:
            raise _classify_status(status)
        try:
            body = response.json()
        except ValueError as exc:
            raise BadResponse("login response not JSON") from exc
        token = body.get("token") or body.get("key") or body.get("access")
        if not token or not isinstance(token, str):
            raise BadResponse("login response missing token/key/access")
        self._token = token
        return token

    def fetch_maa_text(
        self,
        tenant_slug: str,
        vendor: str,
        version: Optional[str] = None,
    ) -> MaaText:
        url = _join_url(self.server_url, RMC_MAA_TEXT_PATH)
        params = {"vendor": vendor, "tenant": tenant_slug}
        if version:
            params["version"] = version
        try:
            with self._client_factory() as client:
                response = client.get(url, params=params, headers=self._auth_header())
        except httpx.HTTPError as exc:
            raise TransportError(f"transport error: {type(exc).__name__}") from exc
        status = response.status_code
        logger.info(
            "rmc_handshake.fetch_maa_text status=%s tenant_slug=%s vendor=%s",
            status, tenant_slug, vendor,
        )
        if status >= 400:
            raise _classify_status(status)
        body = response.json()
        return MaaText(
            vendor=body.get("vendor", vendor),
            version=body.get("version", ""),
            text=body.get("text", ""),
            is_draft=bool(body.get("is_draft", False)),
            active_version=body.get("active_version", ""),
        )

    def sign_and_submit_maa(
        self,
        tenant_slug: str,
        vendor: str,
        holder: str,
        role: str,
        agreement_version: str,
        body_text: str,
    ) -> MAAReceipt:
        url = _join_url(self.server_url, RMC_MAA_SIGN_PATH)
        payload = {
            "vendor_source": vendor,
            "vendor_account_holder_name": holder,
            "signed_by_role": role,
            "agreement_version": agreement_version,
            # The server constant-time-compares this against its own
            # render. We use ``hmac.compare_digest`` on echoes locally
            # in tests, and trust the server's own constant-time check
            # at runtime.
            "submitted_signature_text": body_text,
        }
        try:
            with self._client_factory() as client:
                response = client.post(
                    url,
                    json=payload,
                    headers={**self._auth_header(), "X-Tenant": tenant_slug},
                )
        except httpx.HTTPError as exc:
            raise TransportError(f"transport error: {type(exc).__name__}") from exc
        status = response.status_code
        # NEVER log body_text — only status + IDs.
        logger.info(
            "rmc_handshake.sign_and_submit_maa status=%s tenant_slug=%s vendor=%s",
            status, tenant_slug, vendor,
        )
        if status >= 400:
            if status == 400:
                raise MaaSignRejected(f"MAA sign rejected status={status}")
            raise _classify_status(status)
        parsed = response.json()
        return MAAReceipt(
            maa_id=int(parsed["maa_id"]),
            signed_at=str(parsed.get("signed_at", "")),
            agreement_version=str(parsed.get("agreement_version", "")),
        )

    def fetch_receipt(self, receipt_id: int) -> Receipt:
        path = f"{RMC_RECEIPT_PATH_PREFIX}{receipt_id}/"
        url = _join_url(self.server_url, path)
        try:
            with self._client_factory() as client:
                response = client.get(url, headers=self._auth_header())
        except httpx.HTTPError as exc:
            raise TransportError(f"transport error: {type(exc).__name__}") from exc
        status = response.status_code
        logger.info("rmc_handshake.fetch_receipt status=%s receipt_id=%s", status, receipt_id)
        if status >= 400:
            raise _classify_status(status)
        body = response.json()
        return Receipt(
            receipt_id=int(body.get("receipt_id", receipt_id)),
            bundle_id=body.get("bundle_id"),
            status=str(body.get("status", "")),
        )

File: companion-tauri/src-tauri/src/rmc_handshake.rs
//! RMC platform handshake (v3.37.0 Agent 4).
//!
//! TARGET: RunMyCampus's OWN `/api/v1/...` endpoints. This module
//! NEVER programmatically logs into any third-party SIS (PowerSchool,
//! Blackbaud, Veracross, Alma, FACTS, Skyward). Vendor extraction
//! lives in `companion-extension/` (the browser extension where the
//! operator's own authenticated tab is the security boundary). The
//! Tauri sibling's only login is to the RMC platform itself.
//!
//! Four steps in a strict order:
//!   1. `login(server_url, email, password) -> SessionToken`
//!      POSTs to `/api/v1/auth/login/` with rustls-tls reqwest; no
//!      cookie jar, just a bearer token. Password is zeroed on Drop.
//!   2. `fetch_maa_text(token, tenant_slug, version) -> MaaText`
//!      GETs `/api/v1/migration/maa/text/` (mirrors the existing
//!      `apps/migration_cloud/companion_receiver.py::maa_text_view`).
//!   3. `sign_and_submit_maa(token, tenant_slug, body, attestation)
//!      -> MAAReceipt`
//!      POSTs `/api/v1/migration/maa/sign/` echoing the verbatim body
//!      so the server can constant-time-compare against its own
//!      rendered active-version body.
//!   4. `fetch_receipt(token, receipt_id) -> Receipt`
//!      GETs `/api/v1/migration/receipts/<id>/`.
//!
//! Secret hygiene:
//!   * `Credentials` (email+password) and `SessionToken` both impl
//!     `Drop` to zeroize their inner bytes.
//!   * No `log::*` call EVER prints password, token, or signature_text.
//!     We log: server URL host, tenant slug, status code, receipt id.
//!   * Token comparison uses `subtle::ConstantTimeEq`.

use serde::{Deserialize, Serialize};
use std::fmt;
use std::time::Duration;
use subtle::ConstantTimeEq;
use thiserror::Error;
use zeroize::{Zeroize, ZeroizeOnDrop};

const RMC_LOGIN_PATH: &str = "/api/v1/auth/login/";
const RMC_MAA_TEXT_PATH: &str = "/api/v1/migration/maa/text/";
const RMC_MAA_SIGN_PATH: &str = "/api/v1/migration/maa/sign/";
const RMC_RECEIPT_PATH_PREFIX: &str = "/api/v1/migration/receipts/";

const DEFAULT_TIMEOUT_SECS: u64 = 30;

#[derive(Debug, Error)]
pub enum HandshakeError {
    #[error("invalid server URL")]
    BadServerUrl,
    #[error("network transport error")]
    Transport,
    #[error("authentication failed (401)")]
    Unauthorized,
    #[error("forbidden (403)")]
    Forbidden,
    #[error("not found (404)")]
    NotFound,
    #[error("server returned status {0}")]
    HttpStatus(u16),
    #[error("response decode failed")]
    BadResponse,
    #[error("MAA fetch failed")]
    MaaFetchFailed,
    #[error("MAA signing rejected by server")]
    MaaSignRejected,
}

/// Operator credentials for the RMC platform login only.
///
/// `Zeroize` + `ZeroizeOnDrop` so the password bytes are overwritten
/// the moment the struct goes out of scope.
#[derive(Zeroize, ZeroizeOnDrop)]
pub struct Credentials {
    pub email: String,
    pub password: String,
}

impl fmt::Debug for Credentials {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        f.debug_struct("Credentials")
            .field("email", &"<redacted>")
            .field("password", &"<redacted>")
            .finish()
    }
}

/// Bearer token returned by `/api/v1/auth/login/`. Zeroized on Drop.
#[derive(Zeroize, ZeroizeOnDrop)]
pub struct SessionToken {
    pub bearer: String,
}

impl SessionToken {
    pub fn new(bearer: String) -> Self {
        Self { bearer }
    }

    /// Constant-time equality so a token-comparison loop never leaks
    /// information by short-circuiting.
    pub fn ct_eq(&self, other: &SessionToken) -> bool {
        self.bearer
            .as_bytes()
            .ct_eq(other.bearer.as_bytes())
            .into()
    }

    /// Authorization header value. The token itself is not logged;
    /// callers must only pass this directly into the HTTP client.
    pub fn authorization_header(&self) -> String {
        format!("Bearer {}", self.bearer)
    }
}

impl fmt::Debug for SessionToken {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        f.debug_struct("SessionToken")
            .field("bearer", &"<redacted>")
            .finish()
    }
}

#[derive(Debug, Clone, Deserialize)]
pub struct MaaText {
    pub vendor: String,
    pub version: String,
    pub text: String,
    pub is_draft: bool,
    pub active_version: String,
}

#[derive(Debug, Clone, Deserialize)]
pub struct MAAReceipt {
    pub maa_id: i64,
    pub signed_at: String,
    pub agreement_version: String,
}

#[derive(Debug, Clone, Deserialize)]
pub struct Receipt {
    pub receipt_id: i64,
    pub bundle_id: Option<i64>,
    pub status: String,
}

#[derive(Debug, Serialize)]
struct LoginRequest<'a> {
    email: &'a str,
    password: &'a str,
}

#[derive(Debug, Deserialize)]
struct LoginResponse {
    /// Server returns ``{"token": "<bearer>"}``; many Django DRF
    /// projects use ``key`` or ``access`` — accept all three to ease
    /// integration. Exactly one MUST be present.
    #[serde(default)]
    token: Option<String>,
    #[serde(default)]
    key: Option<String>,
    #[serde(default)]
    access: Option<String>,
}

#[derive(Debug, Serialize)]
struct MaaSignRequest<'a> {
    vendor_source: &'a str,
    vendor_account_holder_name: &'a str,
    signed_by_role: &'a str,
    agreement_version: &'a str,
    /// Verbatim body the client rendered, so the server can compare
    /// byte-for-byte against its own render. Mirrors the
    /// ``submitted_signature_text`` contract in companion_receiver.py.
    submitted_signature_text: &'a str,
}

/// Build an HTTP client with rustls-TLS, no cookie store, sensible
/// timeouts. We do NOT enable a cookie jar — the RMC platform issues a
/// bearer token and we carry it explicitly on every request.
pub fn build_client() -> Result<reqwest::Client, HandshakeError> {
    reqwest::Client::builder()
        .use_rustls_tls()
        .cookie_store(false)
        .https_only(false) // dev may target http://localhost
        .timeout(Duration::from_secs(DEFAULT_TIMEOUT_SECS))
        .user_agent("RunMyCampus-Companion-Tauri/3.37.0")
        .build()
        .map_err(|_| HandshakeError::Transport)
}

fn join_url(base: &str, path: &str) -> Result<String, HandshakeError> {
    let trimmed = base.trim_end_matches('/');
    if trimmed.is_empty() || !(trimmed.starts_with("http://") || trimmed.starts_with("https://")) {
        return Err(HandshakeError::BadServerUrl);
    }
    Ok(format!("{trimmed}{path}"))
}

/// Map HTTP status to a typed error. Used by every endpoint wrapper.
fn classify_status(status: u16) -> HandshakeError {
    match status {
        401 => HandshakeError::Unauthorized,
        403 => HandshakeError::Forbidden,
        404 => HandshakeError::NotFound,
        s => HandshakeError::HttpStatus(s),
    }
}

/// Step 1: login to the RMC platform. Returns a `SessionToken` carrying
/// the bearer string. The plaintext password is zeroed on Drop.
pub async fn login(
    server_url: &str,
    creds: Credentials,
) -> Result<SessionToken, HandshakeError> {
    let url = join_url(server_url, RMC_LOGIN_PATH)?;
    let client = build_client()?;
    let body = LoginRequest {
        email: &creds.email,
        password: &creds.password,
    };
    let response = client
        .post(&url)
        .json(&body)
        .send()
        .await
        .map_err(|_| HandshakeError::Transport)?;
    let status = response.status().as_u16();
    // NEVER log the bearer or password. ID + status only.
    log::info!("rmc_handshake.login status={} email_hash_prefix=<sha>", status);
    if !response.status().is_success() {
        return Err(classify_status(status));
    }
    let parsed: LoginResponse = response
        .json()
        .await
        .map_err(|_| HandshakeError::BadResponse)?;
    let bearer = parsed
        .token
        .or(parsed.key)
        .or(parsed.access)
        .ok_or(HandshakeError::BadResponse)?;
    drop(creds); // Explicit zeroize trigger via ZeroizeOnDrop.
    Ok(SessionToken::new(bearer))
}

/// Step 2: fetch the verbatim MAA body the operator will sign.
pub async fn fetch_maa_text(
    server_url: &str,
    token: &SessionToken,
    tenant_slug: &str,
    vendor: &str,
    version: Option<&str>,
) -> Result<MaaText, HandshakeError> {
    let url = join_url(server_url, RMC_MAA_TEXT_PATH)?;
    let client = build_client()?;
    let mut query: Vec<(&str, &str)> = vec![("vendor", vendor), ("tenant", tenant_slug)];
    if let Some(v) = version {
        query.push(("version", v));
    }
    let response = client
        .get(&url)
        .header("Authorization", token.authorization_header())
        .query(&query)
        .send()
        .await
        .map_err(|_| HandshakeError::Transport)?;
    let status = response.status().as_u16();
    log::info!(
        "rmc_handshake.fetch_maa_text status={} tenant_slug={} vendor={}",
        status, tenant_slug, vendor
    );
    if !response.status().is_success() {
        return Err(classify_status(status));
    }
    let text: MaaText = response
        .json()
        .await
        .map_err(|_| HandshakeError::MaaFetchFailed)?;
    Ok(text)
}

/// Step 3: sign and submit the MAA. The `body` argument MUST be the
/// verbatim text fetched in step 2 — the server constant-time-compares
/// our submitted body against its own render and rejects on drift.
pub async fn sign_and_submit_maa(
    server_url: &str,
    token: &SessionToken,
    tenant_slug: &str,
    vendor: &str,
    holder: &str,
    role: &str,
    agreement_version: &str,
    body: &str,
) -> Result<MAAReceipt, HandshakeError> {
    let url = join_url(server_url, RMC_MAA_SIGN_PATH)?;
    let client = build_client()?;
    let req = MaaSignRequest {
        vendor_source: vendor,
        vendor_account_holder_name: holder,
        signed_by_role: role,
        agreement_version,
        submitted_signature_text: body,
    };
    let response = client
        .post(&url)
        .header("Authorization", token.authorization_header())
        .header("X-Tenant", tenant_slug)
        .json(&req)
        .send()
        .await
        .map_err(|_| HandshakeError::Transport)?;
    let status = response.status().as_u16();
    // NEVER log the body itself — only status + sha prefix would be OK
    // here, but we keep it strictly to status + IDs to avoid any
    // accidental leak of signature_text.
    log::info!(
        "rmc_handshake.sign_and_submit_maa status={} tenant_slug={} vendor={}",
        status, tenant_slug, vendor
    );
    if !response.status().is_success() {
        if status == 400 {
            return Err(HandshakeError::MaaSignRejected);
        }
        return Err(classify_status(status));
    }
    let receipt: MAAReceipt = response
        .json()
        .await
        .map_err(|_| HandshakeError::BadResponse)?;
    Ok(receipt)
}

/// Step 4: fetch a receipt by id (after upload). The receipt forms
/// the audit chain root.
pub async fn fetch_receipt(
    server_url: &str,
    token: &SessionToken,
    receipt_id: i64,
) -> Result<Receipt, HandshakeError> {
    let path = format!("{RMC_RECEIPT_PATH_PREFIX}{receipt_id}/");
    let url = join_url(server_url, &path)?;
    let client = build_client()?;
    let response = client
        .get(&url)
        .header("Authorization", token.authorization_header())
        .send()
        .await
        .map_err(|_| HandshakeError::Transport)?;
    let status = response.status().as_u16();
    log::info!("rmc_handshake.fetch_receipt status={} receipt_id={}", status, receipt_id);
    if !response.status().is_success() {
        return Err(classify_status(status));
    }
    let receipt: Receipt = response
        .json()
        .await
        .map_err(|_| HandshakeError::BadResponse)?;
    Ok(receipt)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn join_url_rejects_missing_scheme() {
        assert!(matches!(
            join_url("example.com", RMC_LOGIN_PATH),
            Err(HandshakeError::BadServerUrl)
        ));
    }

    #[test]
    fn join_url_trims_trailing_slash() {
        let u = join_url("https://rmc.example.com/", RMC_LOGIN_PATH).unwrap();
        assert_eq!(u, "https://rmc.example.com/api/v1/auth/login/");
    }

    #[test]
    fn classify_status_401_unauthorized() {
        assert!(matches!(classify_status(401), HandshakeError::Unauthorized));
    }

    #[test]
    fn classify_status_403_forbidden() {
        assert!(matches!(classify_status(403), HandshakeError::Forbidden));
    }

    #[test]
    fn classify_status_404_not_found() {
        assert!(matches!(classify_status(404), HandshakeError::NotFound));
    }

    #[test]
    fn credentials_debug_redacts() {
        let c = Credentials {
            email: "op@example.com".into(),
            password: "supersecret".into(),
        };
        let s = format!("{:?}", c);
        assert!(!s.contains("supersecret"));
        assert!(s.contains("<redacted>"));
    }

    #[test]
    fn session_token_debug_redacts() {
        let t = SessionToken::new("abc.def.ghi".into());
        let s = format!("{:?}", t);
        assert!(!s.contains("abc.def.ghi"));
        assert!(s.contains("<redacted>"));
    }

    #[test]
    fn session_token_ct_eq_matches() {
        let a = SessionToken::new("same".into());
        let b = SessionToken::new("same".into());
        let c = SessionToken::new("diff".into());
        assert!(a.ct_eq(&b));
        assert!(!a.ct_eq(&c));
    }

    #[test]
    fn token_zeroized_on_drop() {
        // Construct, capture pointer to inner bytes (via len), drop,
        // then assert the zeroize trait was applied (compile-time
        // assertion via trait bound — the macro generates Drop+Zeroize).
        fn assert_zeroize<T: zeroize::ZeroizeOnDrop>(_: &T) {}
        let t = SessionToken::new("token".into());
        assert_zeroize(&t);
        drop(t);
    }
}
